microcosm/annotations.py

```diff
@@ -0,0 +1,106 @@
+"""
+Configuration annotations.
+
+"""
+from distutils.util import strtobool
+
+
+def boolean(value):
+    """
+    Annotate a value as boolean.
+
+    """
+    if value == "":
+        return False
+    return strtobool(value)
+
+
+def required(value):
+    """
+    Annotate a value as required.
+
+    """
+    if value is None:
+        raise Exception("Missing required value")
+    return value
+
+
+def must_override(metadata, value):
+    """
+    Annotate a value as required outside of unit testing.
+
+    """
+    if metadata.testing:
+        return value
+
+    return required(value)
+
+
+def evaluate(metadata, annotations, config):
+    """
+    Evaluate a set of annotations against a `Configuration`.
+
+    """
+    keys = set(annotations.keys()) | set(config.keys())
+    for key in keys:
+        annotation = annotations.get(key)
+        target = config.get(key)
+
+        try:
+            value = evaluate_one(metadata, key, annotation, target)
+        except Exception as error:
+            print(f"Unable to load configuration for '{key}': {target}")  # noqa
+            raise
+
+        config[key] = value
+
+    return config
+
+
+def split(dct):
+    """
+    Split inputs into data and annotations.
+
+    """
+    data, annotations = dict(), dict()
+
+    for key, value in dct.items():
+        if isinstance(value, dict):
+            data[key], annotations[key] = split(value)
+        elif callable(value):
+            annotations[key] = value
+        else:
+            data[key] = value
+
+    return data, annotations
+
+
+def evaluate_one(metadata, key, annotation, target):
+    """
+    Evaluate a single annotation against its target.
+
+    """
+    if annotation is None:
+        return target
+
+    if isinstance(annotation, dict):
+        if not isinstance(target, dict):
+            raise Exception("Expected dictionary")
+
+        subkeys = set(annotation.keys()) | set(target.keys())
+        return {
+            subkey: evaluate_one(
+                metadata,
+                subkey,
+                annotation.get(subkey),
+                target.get(subkey),
+            )
+            for subkey in subkeys
+        }
+    elif annotation is None:
+        return target
+    else:
+        try:
```

Comment:
This is a little much, what's the value of not just standardizing on the signature for an annotation to be `def an_annotation(value, metadata=None)` and being consistent?

Reply:
That won't work for any of the built-in types:

```
>>> int("4", metadata=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'metadata' is an invalid keyword argument for this function
```

However, as I think more about composing optional/required, typing, and defaults, I'm less inclined to try to support built-ins directly.